crates/jstz_core/src/kv/transaction.rs

```diff
@@ -43,258 +40,353 @@ use std::rc::Rc;
-        }
-        Ok(())
+    pub fn remove(&mut self, key: Key) -> Result<()> {
+        self.csnapshot_remove(key)
     }
 
     /// Returns the given key's corresponding entry in the transactional
     /// snapshot for in-place manipulation.
     pub fn entry<'a, 'b, V>(
         &'a mut self,
         rt: &impl Runtime,
-        key: OwnedPath,
+        key: Key,
     ) -> Result<Entry<'b, V>>
     where
         V: Value + DeserializeOwned,
         'a: 'b,
     {
-        self.lookup::<V>(rt, key.clone())?;
-
-        match self.snapshot.entry(key) {
-            btree_map::Entry::Vacant(inner) => Ok(Entry::Vacant(VacantEntry::new(inner))),
-            btree_map::Entry::Occupied(inner) => Ok(Entry::Occupied(OccupiedEntry::new(
-                &mut self.remove_set,
-                inner,
-            ))),
+        // A mutable lookup ensures the key is in the current snapshot
+        self.lookup_mut::<V>(rt, key.clone())?;
+
+        let csnapshot_idx = self.csnapshot_idx();
+        // self.csnapshot() inlined to avoid lifetime issue
+        let csnapshot = self
+            .stack
+            .last_mut()
+            .ok_or(KvError::TransactionStackEmpty)?;
+
+        match csnapshot.insert_edits.entry(key) {
+            btree_map::Entry::Vacant(inner) => {
+                Ok(Entry::vacant(inner, &mut self.lookup_map, csnapshot_idx))
+            }
+            btree_map::Entry::Occupied(inner) => {
+                Ok(Entry::occupied(inner, &mut csnapshot.remove_edits))
+            }
         }
     }
 
-    /// Begins a new transaction
-    pub fn begin(parent: Rc<RefCell<Transaction>>) -> Rc<RefCell<Transaction>> {
-        let child = Rc::new(RefCell::new(Transaction::new()));
-        child.deref().borrow_mut().parent = Some(parent);
-        Rc::clone(&child)
+    /// Begin a transaction.
+    pub fn begin(&mut self) {
+        self.stack.push(Snapshot::default())
     }
 
-    /// Commit a transaction. Returns `true` if the transaction
-    /// was successfully committed to the persistent key-value store.
-    pub fn commit<V>(&mut self, rt: &mut impl Runtime) -> Result<bool>
-    where
-        V: Value,
-    {
-        // Perform deletions
-        match &self.parent {
-            Some(parent) => {
-                for key in &self.remove_set.clone() {
-                    parent.deref().borrow_mut().snapshot.remove(key);
-                    parent.deref().borrow_mut().remove_set.insert(key.clone());
-                }
+    /// Commit a transaction.
+    pub fn commit(&mut self, rt: &mut impl Runtime) -> Result<()> {
+        let curr_ctxt = self.stack.pop().ok_or(KvError::TransactionStackEmpty)?;
+
+        // Following the `.pop`, `prev_idx` is the index of prev_idx (if it exists)
+        let prev_idx = self.csnapshot_idx();
+
+        if let Some(prev_ctxt) = self.stack.last_mut() {
+            // TODO: These clones are probably uncessary since the entry of btree will always be occupied.
```

Comment:
can you please check and remove if it is possible

Reply:
Ditto, will be fixed in a future PR -- fix may require non-trivial changes